vllm/v1/worker/gpu_model_runner.py

```diff
@@ -445,16 +448,42 @@ def execute_model(
         # TODO(woosuk): Avoid the copy. Optimize.
         self.inputs_embeds[:num_scheduled_tokens].copy_(inputs_embeds)
 
+        def should_execute_model(role: RoleOption, prefill_step: bool) -> bool:
+            if role == "both":
+                return True
+            if role == "prefill":
+                return prefill_step
+            if role == "decode":
+                return not prefill_step
+            return False
+
         # Run the decoder.
         # Use persistent buffers for CUDA graphs.
         with set_forward_context(attn_metadata):
-            hidden_states = self.model(
-                input_ids=None,
-                positions=self.positions[:num_input_tokens],
-                kv_caches=self.kv_caches,
-                attn_metadata=None,
-                inputs_embeds=self.inputs_embeds[:num_input_tokens],
-            )
+            role = get_role()
+            # For the first step in decode, we want to receive hidden_state and kv_cache
+            #[hack] Probably not a good way to check if it's prefill_step
+            prefill_step = torch.sum(self.kv_caches[0]) == 0
+            if role == "decode" and prefill_step:
+                hidden_states = torch.empty_like(self.inputs_embeds[:num_input_tokens])
+                dist.recv(hidden_states, src=0)
+                for i in range(len(self.kv_caches)):
+                    dist.recv(self.kv_caches[i], src=0)
+
+            if should_execute_model(role, prefill_step):
+                hidden_states = self.model(
+                    input_ids=None,
+                    positions=self.positions[:num_input_tokens],
+                    kv_caches=self.kv_caches,
+                    attn_metadata=None,
+                    inputs_embeds=self.inputs_embeds[:num_input_tokens],
+                )
+
+            if role == "prefill" and prefill_step:
+                dist.send(hidden_states, dst=1)
+                for i in range(len(self.kv_caches)):
+                    dist.send(self.kv_caches[i], dst=1)
```

Comment:
Currently you send the entire kvcache. Next step we want to do is just sending the kvcache of specific requests' block ids which you can find in `scheduler_output.scheduled_new_reqs` or `scheduler_output.scheduled_resumed_reqs`